accord-core/src/main/java/accord/local/Command.java

```diff
@@ -274,99 +289,110 @@ public boolean commit(Txn txn, Key homeKey, Key progressKey, Timestamp executeAt
             });
         }
 
-        if (waitingOnCommit.isEmpty())
-        {
-            waitingOnCommit = null;
-            if (waitingOnApply.isEmpty())
-                waitingOnApply = null;
-        }
 
-        boolean isProgressShard = progressKey != null && handles(txnId.epoch, progressKey);
-        commandStore.progressLog().commit(txnId, isProgressShard, isProgressShard && progressKey.equals(homeKey));
+        boolean isProgressShard = progressKey != null && handles(txnId().epoch, progressKey);
+        commandStore().progressLog().commit(txnId(), isProgressShard, isProgressShard && progressKey.equals(homeKey));
 
-        maybeExecute(false);
-        listeners.forEach(this);
         return true;
     }
 
+    public Future<Void> commitAndBeginExecution(Txn txn, Key homeKey, Key progressKey, Timestamp executeAt, Deps deps)
+    {
+        if (!commit(txn, homeKey, progressKey, executeAt, deps))
+            return Write.SUCCESS;
+
+        return maybeExecute(true);
+    }
+
+    // TODO (now): commitInvalidate may need to update cfks _if_ possible
     public boolean commitInvalidate()
     {
         if (hasBeen(Committed))
         {
+            logger.trace("{}: skipping commit invalidated - already committed ({})", txnId(), status());
             if (!hasBeen(Invalidated))
-                commandStore.agent().onInconsistentTimestamp(this, Timestamp.NONE, executeAt);
+                commandStore().agent().onInconsistentTimestamp(this, Timestamp.NONE, executeAt());
 
             return false;
         }
 
-        status = Invalidated;
+        status(Invalidated);
 
-        boolean isProgressShard = progressKey != null && handles(txnId.epoch, progressKey);
-        commandStore.progressLog().invalidate(txnId, isProgressShard, isProgressShard && progressKey.equals(homeKey));
+        boolean isProgressShard = progressKey() != null && handles(txnId().epoch, progressKey());
+        commandStore().progressLog().invalidate(txnId(), isProgressShard, isProgressShard && progressKey().equals(homeKey()));
+        logger.trace("{}: committed invalidated", txnId());
 
-        listeners.forEach(this);
+        notifyListeners();
         return true;
     }
 
-    public boolean apply(Txn txn, Key homeKey, Key progressKey, Timestamp executeAt, Deps deps, Writes writes, Result result)
+    public Future<Void> apply(Txn txn, Key homeKey, Key progressKey, Timestamp executeAt, Deps deps, Writes writes, Result result)
```

Comment:
Reply from Blake in JIRA

So I don't think your example timeline is possible, since Command methods like commit are idempotent and the command store is single threaded. T2 would be a noop. However, I don't think there's anything to prevent maybeExecute to fire off duplicate writes in the time between Command#apply dispatches writes and the time Command#postApply sets the status to Applied in Command#postApply., and the same goes for Command#read. I'm going to think about this a bit, since it depends on the concurrency of the implementation I'm not sure if it would be better to handle this in the library, or in the implementation....

Reply:
It seems totally reasonable to introduce a new status `Applying` or something, that avoids duplicated work.